CDP4Reporting/DataSource/ReportingDataSourceNodeTreeGenerator.cs

```diff
@@ -0,0 +1,160 @@
+
+    using CDP4Common.EngineeringModelData;
+    using CDP4Common.SiteDirectoryData;
+
+    /// <summary>
+    /// A builder that builds a tree of <see cref="IEnumerable{ReportingDataSourceNode{T}}"/> instances
+    /// </summary>
+    /// <typeparam name="T">
+    /// The <see cref="ReportingDataSourceRow"/> representing the data source rows.
+    /// </typeparam>
+    internal class ReportingDataSourceNodeTreeGenerator<T> where T : ReportingDataSourceRow, new()
+    {
+        public IEnumerable<ReportingDataSourceNode<T>> Generate(CategoryHierarchy categoryHierarchy, List<NestedElement> nestedElements)
+        {
+            var topElement = nestedElements.First(ne => ne.IsRootElement);
+            return this.GetDataSourceNodes(topElement, categoryHierarchy, nestedElements, null);
+        }
+
+        /// <summary>
+        /// Gets the tree of <see cref="ReportingDataSourceNode{T}"/>s. This method can call itself recursively.
+        /// </summary>
+        /// <param name="nestedElement">The <see cref="NestedElement"/></param>
+        /// <param name="categoryHierarchy">The <see cref="CategoryHierarchy"/></param>
+        /// <param name="nestedElements">The <see cref="List{NestedElement{>}"/>s</param>
+        /// <param name="parentNode">The <see cref=" ReportingDataSourceNode{T}"/> that is the parent of new nodes.</param>
+        /// <returns>An <see cref="IEnumerable{ReportingDataSourceNode{T}}"/></returns>
+        private IEnumerable<ReportingDataSourceNode<T>> GetDataSourceNodes(NestedElement nestedElement, CategoryHierarchy categoryHierarchy, List<NestedElement> nestedElements, ReportingDataSourceNode<T> parentNode)
+        {
+            var resultNodes = new List<ReportingDataSourceNode<T>>();
+            ReportingDataSourceNode<T> newNode = null;
+            var searchCategory = categoryHierarchy;
+
+            if (this.HasCategory(nestedElement, categoryHierarchy.Category))
+            {
+                newNode = new ReportingDataSourceNode<T>(categoryHierarchy, nestedElement, parentNode);
+                parentNode?.Children.Add(newNode);
+                resultNodes.Add(newNode);
+            }
+            else if ((parentNode?.HadCategoryUpTree(categoryHierarchy.Category) ?? false) && categoryHierarchy.Child != null && this.HasCategory(nestedElement, categoryHierarchy.Child.Category))
+            {
+                //if parent was found somewhere up the tree
+                searchCategory = categoryHierarchy.Child;
+
+                newNode = new ReportingDataSourceNode<T>(searchCategory, nestedElement, parentNode);
+                parentNode?.Children.Add(newNode);
+                resultNodes.Add(newNode);
+            }
+
+            var children = this.GetChildren(nestedElement, nestedElements).ToList();
+
+            foreach (var child in children)
+            {
+                var nodes = this.GetDataSourceNodes(child, searchCategory, nestedElements, newNode ?? parentNode).ToArray();
```

Comment:
> 
> 
> `newNode ?? parentNode`: this seems like it will cause the structure to no longer match the actual "product tree" structure: the children will not be actually children of the parent, just nodes in the parent's subtree that match the next category in the hierarchy. In most cases this will probably not cause any problem, but as an user I would expect the parent/child relationships to match the model; as an example, if you were to want to have a computed parameter making an average of values from children, this new approach will produce a different result that expected.

True, but this has to do with the way of thinking from nested object structure to tabular data. We only want the data from the deepest nested category level. So averages and sums will then be available in the reporting designer and not in the DataSource selection.

Reply:
Fair point, although initially it seemed like a good idea to have this flexibility in the data source as well.

I propose we get rid of the `GetChildren` (possibly `GetSibling` too?) methods on the `ReportingDataSourceParameter` class - this would also lead to some other collateral simplification in the node class.